doc/man1/genpkey.pod

```diff
@@ -160,91 +156,92 @@ If set the key is restricted and B<len> specifies the minimum salt length.
-and 2048 bit group with 256 bit subgroup as mentioned in RFC5114 sections
-2.1, 2.2 and 2.3 respectively.
+The digest to use during parameter generation. Must be one of sha1, sha224 or
+sha256. If set then the number of bits in q will match the output size of the
+specified digest and the B<dsa_paramgen_q_bits> parameter will be ignored. If
+not set then a digest will be used that gives an output matching the number of
+bits in q, i.e. sha1 if q length is 160, sha224 if it 224 or sha256 if it is
+256.
 
 =back
 
-=head1 EC PARAMETER GENERATION OPTIONS
-
-The EC parameter generation options below can also
-be supplied as EC key generation options. This can (for example) generate a
-key from a named curve without the need to use an explicit parameter file.
+=head2 DH Parameter Generation Options
 
 =over 4
 
-=item B<ec_paramgen_curve:curve>
+=item B<dh_paramgen_prime_len:numbits>
 
-The EC curve to use. OpenSSL supports NIST curve names such as "P-256".
+The number of bits in the prime parameter B<p>. The default is 1024.
 
-=item B<ec_param_enc:encoding>
+=item B<dh_paramgen_subprime_len:numbits>
 
-The encoding to use for parameters. The "encoding" parameter must be either
-"named_curve" or "explicit".
+The number of bits in the sub prime parameter B<q>. The default is 256 if the
+prime is at least 2048 bits long or 160 otherwise. Only relevant if used in
+conjunction with the B<dh_paramgen_type> option to generate X9.42 DH parameters.
 
-=back
-
-=head1 GOST2001 KEY GENERATION AND PARAMETER OPTIONS
-
-Gost 2001 support is not enabled by default. To enable this algorithm,
-one should load the ccgost engine in the OpenSSL configuration file.
-See README.gost file in the engines/ccgost directory of the source
-distribution for more details.
+=item B<dh_paramgen_generator:value>
 
-Use of a parameter file for the GOST R 34.10 algorithm is optional.
-Parameters can be specified during key generation directly as well as
-during generation of parameter file.
+The value to use for the generator B<g>. The default is 2.
 
-=over 4
+=item B<dh_paramgen_type:value>
 
-=item B<paramset:name>
+The type of DH parameters to generate. Use 0 for PKCS#3 DH and 1 for X9.42 DH.
+The default is 0.
 
-Specifies GOST R 34.10-2001 parameter set according to RFC 4357.
-Parameter set can be specified using abbreviated name, object short name or
-numeric OID. Following parameter sets are supported:
+=item B<dh_rfc5114:num>
 
-  paramset   OID               Usage
-  A          1.2.643.2.2.35.1  Signature
-  B          1.2.643.2.2.35.2  Signature
-  C          1.2.643.2.2.35.3  Signature
-  XA         1.2.643.2.2.36.0  Key exchange
-  XB         1.2.643.2.2.36.1  Key exchange
-  test       1.2.643.2.2.35.0  Test purposes
+If this option is set then the appropriate RFC5114 parameters are used
+instead of generating new parameters. The value B<num> can take the
+values 1, 2 or 3 corresponding to RFC5114 DH parameters consisting of
```

Comment:
What a hack.  Shouldn't we also take 160 224 256?

Reply:
I'm not really convinced that makes things a lot better.

Reply:
Sorry, for the unrelated question.
Aren't the 5114 parameters DSA parameters ?

Reply:
Err no? RFC 5114 = "Additional Diffie-Hellman Groups for Use with IETF Standards"

Reply:
Just curious, because they have these small subgroups that usually DSA needs.

Reply:
The X9.42 variant of DH has them too.

Reply:
Sorry, again for completely unrelated chat...
Would'nt they be usable as pre-defined DSA parameters?

Reply:
I *think* it should be theoretically possible. We in fact do have support for converting DSA parameters into DH parameters (e.g. see the DSA_dup_DH() function). There are warnings in the docs about possible small subgroup attacks in this case because historically we didn't have support for "q" in DH. However now that we have support for that (due to X9.42) that may not be an issue any more?? Anyway, AFAICT, we don't have an equivalent DH_dup_DSA() function so it can't, in practice, be done. These specific parameters are only defined for DH use, so I'm not sure if it would be wise to encourage their use in other settings.